Human readable OSError messages

godlygeek opened the pull request:
For several common errnos, include a human-friendly string in the exception string.  For now, only enable this on boards with external SPI flash, because pulling the `uerrno` module in eats up an extra 1kb of flash.

Likewise, this only adds error strings for common exceptions to avoid using flash unnecessarily - others will just show up as the EXXX string, but meaningful descriptions can be added later if needed.

Tested with a Feather M0 Express REPL:

    Adafruit CircuitPython 3.0.0-alpha.6-75-gcb8c159-dirty on 2018-05-14; Adafruit Feather M0 Express with samd21g18
    >>> import uerrno
    >>> for err in dir(uerrno):
    ...     if err.startswith('E'):
    ...         print(OSError(getattr(uerrno, err)))
    ...
    ...
    ...
    [Errno 1] Permission denied
    [Errno 2] No such file/directory
    [Errno 5] Input/output error
    [Errno 9] EBADF
    [Errno 11] EAGAIN
    [Errno 12] ENOMEM
    [Errno 13] Permission denied
    [Errno 17] File exists
    [Errno 19] Unsupported operation
    [Errno 21] EISDIR
    [Errno 22] Invalid argument
    [Errno 95] EOPNOTSUPP
    [Errno 112] EADDRINUSE
    [Errno 113] ECONNABORTED
    [Errno 104] ECONNRESET
    [Errno 105] ENOBUFS
    [Errno 128] ENOTCONN
    [Errno 116] ETIMEDOUT
    [Errno 111] ECONNREFUSED
    [Errno 118] EHOSTUNREACH
    [Errno 120] EALREADY
    [Errno 119] EINPROGRESS

Signed-off-by: Matt Wozniski <mwozniski@bloomberg.net>

dhalbert commented:
The Unix build runs the test - one test is showing unexpected output. May just need to modify the test, or handle that case.
https://travis-ci.org/adafruit/circuitpython/jobs/378877220
```
### Expected
b"<class 'int'>\n[Errno  ] EIO\n9999\n"
### Actual
b"<class 'int'>\n[Errno  error\n9999\n"
FAIL  basics/errno1.py
```

godlygeek commented:
The test did need to be updated; I've done so.  The test was checking the output of:

    str(OSError(uerrno.EIO))

Previously it would return something like `'[Errno 5] EIO'`, but because the "5" is platform specific, the test was checking only the first 7 characters (`[Errno `) and the last 5 (`] EIO`).

We deliberately changed the message to `'[Errno 5] Input/output error'` and so looking at only the last 5 was no longer correct.  The test is now updated to look for everything after the first `]` instead of only the last 5 characters, and to expect "Input/output error" instead of "EIO" as the description.  I also added a test for one that we didn't give a human-readable description to, to show that we still stringify the error name in that case.
